<a href="https://colab.research.google.com/github/bacchus00/tareas-ds/blob/main/Competencia_1_NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Competencia 1 - CC6205 Natural Language Processing 📚**

**Integrantes:** Ricardo García, Nicolás Canales, Melanie Marquez, Franc Zautzik

**Usuario del equipo en CodaLab:** FraNiMeRi

**Fecha límite de entrega 📆:** Jueves 14 de Abril.

**Tiempo estimado de dedicación:** 20 hrs

## **Objetivo**

En esta tarea grupal participarán de una competencia estilo [Kaggle](https://www.kaggle.com/) pero utilizando la página [CodaLab](https://codalab.org/). El objetivo es trabajar en la clasificación de tweets  según su intensidad de emoción, esto corresponde a la task de clasificación de texto. 

Tendrán a su disposición 4 datasets de tweets con distintas emociones: `anger`, `fear`, `sadness` y `joy`. Deberán crear un clasificador para cada uno de estos datasets que indique la intensidad de dicha emoción en sus tweets (`low`, `medium`, `high`). 

## **Instrucciones**



- La competencia consiste en resolver 4 problemas de clasificación **distintos**, cada uno con tres clases posibles. Por cada problema deberán crear un clasificador distinto. La evaluación de la competencia se realiza en base a 3 métricas: `AUC`, `Kappa` y `Accuracy`. Los mejores puntajes en cada una de estas métricas serán quienes ganen la competencia.

- Para comenzar se les entregará en este notebook la estructura del informe a entregar y el código de un baseline con el cuál pueden comparar los resultados de sus experimentos. Este baseline consiste en la creación de features y una clasificación simple siguiendo el paradigma de Machine Learning empírico. Los puntajes obtenidos por el baseline los pueden visualizar en el link de CodaLab buscando al usuario **cc6205**. Esperamos que los superen fácilmente 😉

- Para participar, deben registrarse en CodaLab y luego ingresar a la competencia usando el siguiente [link](https://competitions.codalab.org/competitions/30330?secret_key=b2ed0be7-bf23-42a1-9400-f85fa1b7bae7). 

- Está permitido hacer grupos de máximo 4 alumnos. Cada grupo debe tener un nombre de equipo, para ello en CodaLab pueden dirigirse a settings y luego cambiar el Team Name. Sólo una persona debe administrar la cuenta del grupo y se verificará que no se hayan creado múltiples cuentas por grupo.

- En total pueden hacer un **máximo de 4 submissions**, hagan muchos experimentos probando en el conjunto de test antes de realizar el envío.

- Es importante que hagan varios experimentos incorporando técnicas como [cross-validation](https://es.wikipedia.org/wiki/Validaci%C3%B3n_cruzada#:~:text=La%20validaci%C3%B3n%20cruzada%20o%20cross,datos%20de%20entrenamiento%20y%20prueba.) o [random sampling](https://towardsdatascience.com/the-5-sampling-algorithms-every-data-scientist-need-to-know-43c7bc11d17c) antes de enviar sus predicciones a CodaLab, ya que les puede dar un mejor indicio del nivel de generalización de sus modelos. 

- Asegúrense que la distribución de las clases sea balanceada en las particiones de training y testing debido a que existe un desbalanceo. 

- Verificar que el formato de la submission coincida con el de la competencia. De lo contrario, se les será evaluado incorrectamente ya que el Script de evaluación espera como input dicho formato. En el código de las métricas pueden verificar cómo son los inputs.

- No se limiten a los contenidos vistos ni a scikit ni a este baseline. No tienen restricciones entre utilizar Deep Learning o Machine Learning empírico. Si reutilizan gran cantidad de código de alguna página por favor mostrar la referencia en su código. ¡Usen todo su conocimiento e ingenio en mejorar sus sistemas para poder ganar la competencia! 

- **Es requisito entregar el reporte con el código y haber participado en la competencia para ser evaluado. Un código sin reporte o un reporte sin código serán evaluados con la nota mínima.**

- Todas las dudas escríbanlas en el canal de Discord de competencias. Los emails que lleguen al equipo docente serán remitidos a ese medio.

**Importante**: Recuerden poner su nombre y el de su usuario o de equipo (en caso de que aplique) en el reporte. NO serán evaluados Notebooks sin nombre.

----------------------------------------

## **Reporte a entregar**

Uno de los puntos claves en la evaluación de esta competencia es elaborar un informe claro y preciso, argumentando las decisiones tomadas al momento de crear sus modelos para que el cuerpo docente pueda entenderlos. La estructura que debe contener es la siguiente:

1.	**Introducción**: Presentar brevemente el problema a resolver, incluyendo la formalización de la task (cómo son los inputs y outputs del problema) y los desafíos que ven al analizar el corpus entregado. (**0.5 puntos**)
2.	**Representaciones**: Describir los atributos y representaciones usadas como entrada de los clasificadores, **recordar** que para entrenar modelos el input debe tener su representación numérica. Si bien, con Bag of Words (**baseline**) ya se comienzan a percibir buenos resultados, pueden mejorar su evaluación agregando más atributos y representaciones diseñadas a mano, sean lo más creativos posible. Más abajo encontrarán una lista de estos posibles atributos que les podrá ser de utilidad. (**1 punto**)
3.	**Algoritmos**: Describir **brevemente** los algoritmos de clasificación usados, tanto si fueron algoritmos ya vistos en clases o bien arquitecturas de Deep Learning. (**0.5 puntos**)
4.	**Métricas de evaluación**: Describir brevemente las métricas utilizadas en la evaluación, indicando qué miden y su interpretación. (**0.5 puntos**)

5.  **Diseño experimental**: Esta es una de las secciones más importantes del reporte. Deben describir minuciosamente los experimentos que realizarán en la siguiente sección. Describir las variables de control que manejarán, algunos ejemplos pueden ser: Los parámetros de los clasificadores, parámetros en las funciones con que procesan los textos y los transforman, parámetros para el cross-validation, particiones de datos utilizadas, etc. En caso que utilicen redes neuronales, ser claros con el conjunto de hiperparámetros que probarán, la decisión en las funciones de optimización, función de pérdida,  regulación, etc. Básicamente explicar qué es lo que veremos en la siguiente sección.
(**1 punto**)

6.	**Experimentos**: Incluyan todo el código de sus experimentos aquí. ¡Es vital haber realizado varios experimentos para sacar una buena nota! (**1.5 puntos**)

7. **Resultados**: Comparar los resultados obtenidos utilizando diferentes algoritmos y representaciones.  Pueden mostrar los resultados sobre la partición de validación en caso que la generen o sobre los resultados del conjunto de testing. Mostrar los resultados en alguna tabla, pueden poner aquí también los resultados obtenidos al realizar la submission. (**0.5 puntos**)

8. **Conclusiones**:  Discutir resultados, proponer trabajo futuro. (**0.5 puntos**)

## **Descripción Baseline**

` `
El baseline de la sección **Experimentos** contiene un código básico que resuelve la task y es el modelo subido por **cc6205-baseline** a la competencia. Pueden modificar el código como quieran siempre y cuando no cambien las funciones de las métricas y el formato en que se suben los archivos a la competencia, ya que ese es el formato en que realizamos la evaluación. En concretro, el baseline hace lo siguiente:
` `

- Obtiene los datasets desde el repositorio del curso.
- Divide los datasets en train (datos que están etiquetados) y target set (datos no etiquetados para la competencia). Además, por cada dataset en train, se divide en un conjunto de entrenamiento y uno de prueba. Aquí la mejor práctica sería cambiar el código para obtener un conjunto de entrenamiento, validación y prueba.

- Crea un Pipeline que: 
    - Crea features personalizados para la representación numérica.
    - Transforma los dataset a bag of words (BoW).  
    - Entrena un clasificador usando cada train set.
- Clasifica y evalua el sistema creado usando el test set.
- Clasifica el target set.
- Genera una submission con el target en formato zip en el directorio en donde se está ejecutando el notebook. 

` `

Algunas pistas sobre como mejorar el rendimiento de los sistemas que creen. (Esto tendrá mas sentido cuando vean el código)

- **Vectorizador**: Investigar los modulos de `nltk`, en particular, `TweetTokenizer`, `mark_negation` para reemplazar los tokenizadores. También, el parámetro `ngram_range` (Ojo que el clf naive bayes no debería usarse con n-gramas, ya que rompe el supuesto de independencia). Además, implementar los atributos que crean útiles desde el listado del enunciado. Investigar también el vectorizador tf-idf.

- **Clasificador**: Investigar otros clasificadores más efectivos que naive bayes. Estos deben poder retornar la probabilidad de pertenecia de las clases (ie: implementar la función `predict_proba`).

- **Features**: Recuerden que pueden implementar todas las features que se les ocurra! Aquí les adjuntamos algunos ejemplos:
    -	Word n-grams.
    -	Character n-grams. 
    -	Part-of-speech tags.
    -	Sentiment Lexicons (Lexicon = A set of words with a label or associated value.).
        - Count the number of positive and negative words within a sentence.
        - If the lexicon has associated intensity of feeling (for example in a decimal), then take the average of the intensity of the sentence according to the feeling, the sum, etc.
        -	A good lexicon of sentiment: [Bing Liu](http://www.cs.uic.edu/~liub/FBS/opinion-lexicon-English.rar) 
        - A reference with a lot of [sentiment lexicons](https://medium.com/@datamonsters/sentiment-analysis-tools-overview-part-1-positive-and-negative-words-databases-ae35431a470c). 
    -	The number of elongated words (words with one character repeated more than two times).
    -	The number of words with all characters in uppercase.
    -	The presence and the number of positive or negative emoticons.
    -	The number of individual negations.
    -	The number of contiguous sequences of dots, question marks and exclamation marks.
    -	Word Embeddings: Here are some good ideas on how to use them.
    https://stats.stackexchange.com/questions/221715/apply-word-embeddings-to-entire-document-to-get-a-feature-vector

- **Reducción de dimensionalidad**: También puede serles de ayuda. Referencias [aquí](https://scikit-learn.org/stable/modules/unsupervised_reduction.html).

- Por último, pueden encontrar mas referencias de cómo mejorar sus features, el vectorizador y el clasificador [aquí](https://affectivetweets.cms.waikato.ac.nz/benchmark/).

# **Entregable.**

## **1. Introducción**

## **2. Representaciones**

## **3. Algoritmos**

> Bloc con sangría



    Explique los algoritmos utilizados

## **4. Métricas de Evaluación**

- AUC: ...
- Kappa: ...
- Accuracy: ...


## **5. Diseño experimental**

    Descripción de la metodología 

## **6. Experimentos**

### Importar librerías

In [ ]:
import pandas as pd
import os
import numpy as np
import shutil
import re

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.metrics import confusion_matrix, cohen_kappa_score, classification_report, accuracy_score, roc_auc_score
from sklearn.model_selection import train_test_split
from sklearn.base import BaseEstimator, TransformerMixin




### Definir métodos de evaluación (**NO tocar este código**)

Estas funciones están a cargo de evaluar los resultados de la tarea. No deberían cambiarlas.


In [ ]:
def auc_score(test_set, predicted_set):
    high_predicted = np.array([prediction[2] for prediction in predicted_set])
    medium_predicted = np.array(
        [prediction[1] for prediction in predicted_set])
    low_predicted = np.array([prediction[0] for prediction in predicted_set])
    high_test = np.where(test_set == 'high', 1.0, 0.0)
    medium_test = np.where(test_set == 'medium', 1.0, 0.0)
    low_test = np.where(test_set == 'low', 1.0, 0.0)
    auc_high = roc_auc_score(high_test, high_predicted)
    auc_med = roc_auc_score(medium_test, medium_predicted)
    auc_low = roc_auc_score(low_test, low_predicted)
    auc_w = (low_test.sum() * auc_low + medium_test.sum() * auc_med +
             high_test.sum() * auc_high) / (
                 low_test.sum() + medium_test.sum() + high_test.sum())
    return auc_w


def evaluate(predicted_probabilities, y_test, labels, dataset_name):
    # Importante: al transformar los arreglos de probabilidad a clases,
    # entregar el arreglo de clases aprendido por el clasificador.
    # (que comunmente, es distinto a ['low', 'medium', 'high'])
    predicted_labels = [
        labels[np.argmax(item)] for item in predicted_probabilities
    ]

    print('Confusion Matrix for {}:\n'.format(dataset_name))
    print(
        confusion_matrix(y_test,
                         predicted_labels,
                         labels=['low', 'medium', 'high']))

    print('\nClassification Report:\n')
    print(
        classification_report(y_test,
                              predicted_labels,
                              labels=['low', 'medium', 'high']))
    # Reorder predicted probabilities array.
    labels = labels.tolist()
    
    predicted_probabilities = predicted_probabilities[:, [
        labels.index('low'),
        labels.index('medium'),
        labels.index('high')
    ]]
    
    
    auc = round(auc_score(y_test, predicted_probabilities), 3)
    print("Scores:\n\nAUC: ", auc, end='\t')
    kappa = round(cohen_kappa_score(y_test, predicted_labels), 3)
    print("Kappa:", kappa, end='\t')
    accuracy = round(accuracy_score(y_test, predicted_labels), 3)
    print("Accuracy:", accuracy)
    print('------------------------------------------------------\n')
    return np.array([auc, kappa, accuracy])

### Datos



In [ ]:
# Datasets de entrenamiento.
train = {
    'anger': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/anger-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity']),
    'fear': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/fear-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity']),
    'joy': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/joy-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity']),
    'sadness': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/sadness-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'])
}

# Datasets que deberán predecir para la competencia.
target = {
    'anger': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/anger-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE']),
    'fear': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/fear-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE']),
    'joy': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/joy-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE']),
    'sadness': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/sadness-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE'])
}

In [ ]:
# Ejemplo de algunas filas aleatorias del dataset etiquetado:
train['anger'].sample(5)

In [ ]:
# Ejemplo de algunas filas aleatorias del dataset no etiquetado
target['anger'].sample(5)

### Analizar los datos 

En esta sección analizaremos el balance de los datos. Para ello se imprime la cantidad de tweets de cada dataset agrupados por la intensidad de sentimiento.

In [ ]:
for dataset_name in train:
    print(f'Dataset: {dataset_name} \n', train[dataset_name].groupby(['sentiment_intensity']).size())
    print('----------------------------------------------------------\n')

Dataset: anger 
 sentiment_intensity
high      163
low       161
medium    617
dtype: int64
----------------------------------------------------------

Dataset: fear 
 sentiment_intensity
high      270
low       288
medium    699
dtype: int64
----------------------------------------------------------

Dataset: joy 
 sentiment_intensity
high      195
low       219
medium    488
dtype: int64
----------------------------------------------------------

Dataset: sadness 
 sentiment_intensity
high      197
low       210
medium    453
dtype: int64
----------------------------------------------------------



In [ ]:
sentiments = ['anger','sadness','fear','joy']
sentiment_intensities = ['high','medium','low']
smallest_set = {'anger':161,'sadness':197,'fear':270,'joy':195}
def typeCounterByIntensity():
  results = { }
  for sentiment in sentiments:
    data_frame = train[sentiment]
    for intensity in sentiment_intensities:
      sentiment_df = data_frame[data_frame['sentiment_intensity']==intensity].sample(smallest_set[sentiment],random_state=1)
      cv = CountVectorizer(stop_words='english')
      cv_fit = cv.fit_transform(sentiment_df.tweet)
      df = pd.DataFrame({'token':cv.get_feature_names_out(), ('count'+' '+intensity):cv_fit.toarray().sum(axis=0)})
      df = df.sort_values(('count'+' '+intensity),ascending=False)
      results[sentiment+ ' ' + intensity]=df
  return results

results = typeCounterByIntensity()


In [ ]:
anger_results = results['anger high'].merge(results['anger medium'].merge(results['anger low'],on='token',how='outer'),on='token',how='outer')
anger_results

token  count high  count medium  count low
0           fuming        23.0           3.0        NaN
1            angry        21.0           7.0        2.0
2           people        18.0          14.0        5.0
3             just        15.0          21.0       17.0
4              don        13.0          10.0       12.0
...            ...         ...           ...        ...
2101        heaven         NaN           NaN        1.0
2102         hello         NaN           NaN        1.0
2103  heymanhustle         NaN           NaN        1.0
2104     hialeshia         NaN           NaN        1.0
2105         zumba         NaN           NaN        1.0

[2106 rows x 4 columns]

In [ ]:
sadness_results = results['sadness high'].merge(results['sadness medium'].merge(results['sadness low'],on='token',how='outer'),on='token',how='outer')
sadness_results

token  count high  count medium  count low
0       depressing        28.0           1.0        NaN
1       depression        26.0           3.0        NaN
2              sad        23.0          16.0        2.0
3          sadness        16.0           7.0        NaN
4          unhappy        15.0          10.0        1.0
...            ...         ...           ...        ...
2755  granddesigns         NaN           NaN        1.0
2756         green         NaN           NaN        1.0
2757  griffinlanep         NaN           NaN        1.0
2758       grilled         NaN           NaN        1.0
2759           アニメ         NaN           NaN        1.0

[2760 rows x 4 columns]

In [ ]:
fear_results = results['fear high'].merge(results['fear medium'].merge(results['fear low'],on='token',how='outer'),on='token',how='outer')
fear_results

token  count high  count medium  count low
0           nervous        36.0           3.0        NaN
1         nightmare        26.0           7.0        1.0
2           anxiety        26.0           6.0        1.0
3             panic        24.0           6.0        NaN
4              just        20.0          16.0       19.0
...             ...         ...           ...        ...
3293     hallelujah         NaN           NaN        1.0
3294  hatersbackoff         NaN           NaN        1.0
3295         harder         NaN           NaN        1.0
3296          happy         NaN           NaN        1.0
3297         handle         NaN           NaN        1.0

[3298 rows x 4 columns]

In [ ]:
joy_results = results['joy high'].merge(results['joy medium'].merge(results['joy low'],on='token',how='outer'),on='token',how='outer')
joy_results

token  count high  count medium  count low
0          happy        39.0          11.0        5.0
1           love        36.0           9.0        4.0
2            day        31.0           5.0        6.0
3          smile        24.0          13.0        4.0
4      hilarious        22.0           8.0        1.0
...          ...         ...           ...        ...
2463   generally         NaN           NaN        1.0
2464        gifs         NaN           NaN        1.0
2465       girls         NaN           NaN        1.0
2466         gkn         NaN           NaN        1.0
2467  yungdoujin         NaN           NaN        1.0

[2468 rows x 4 columns]

### Custom Features 

Para crear features personalizadas implementaremos nuestros propios Transformers (estandar de scikit para crear nuevas features entre otras cosas). Para esto:

1. Creamos nuestra clase Transformer extendiendo BaseEstimator y TransformerMixin. En este ejemplo, definiremos `CharsCountTransformer` que cuenta caracteres relevantes ('!', '?', '#', '@') en los tweets.
2. Definimos una función cómo `get_relevant_chars` que opera por cada tweet y retorna un arreglo.
3. Hacemos un override de la función `transform` en donde iteramos por cada tweet, llamamos a la función que hicimos antes y agregamos sus resultados a un arreglo. Finalmente lo retornamos.

Esto nos facilitará el trabajo mas adelante. Una Guia completa de las transformaciones predefinidas en scikit pueden encontrarla [aquí](https://scikit-learn.org/stable/data_transforms.html).



In [ ]:
class CharsCountTransformer(BaseEstimator, TransformerMixin):
    def get_relevant_chars(self, tweet):
        num_hashtags = tweet.count('#')
        num_exclamations = tweet.count('!')
        num_interrogations = tweet.count('?')
        num_at = tweet.count('@')
        return [num_hashtags]

    def transform(self, X, y=None):
        chars = []
        for tweet in X:
            chars.append(self.get_relevant_chars(tweet))

        return np.array(chars)

    def fit(self, X, y=None):
        return self

In [ ]:
# Veamos que sucede si ejecutamos el transformer
sample = train['anger'].sample(5, random_state = 1).tweet
sample_features = CharsCountTransformer().transform(sample)

# Se puede verificar que el conteo de símbolos es consistente con el transformer creado.
print(f'Tweet original: {sample.iloc[0]}')
print(f'Features creados: {sample_features[0]}')

Tweet original: @everycolorbot more like every color looks the same #triggered #colorblind 
Features creados: [2 0 0 1]


In [ ]:
sentiments = ['anger','sadness','fear','joy']
sentiment_intensities = ['high','medium','low']
smallest_set = {'anger':161,'sadness':197,'fear':270,'joy':195}

def charCounterByIntensity():
  for sentiment in sentiments:
    print("\n----"+sentiment+"----")
    data_frame = train[sentiment]
    for intensity in sentiment_intensities:
      sentiment_df = data_frame[data_frame['sentiment_intensity']==intensity].sample(smallest_set[sentiment])
      features = CharsCountTransformer().transform(sentiment_df.tweet)
      total_count = [0,0,0,0]
      for i in range(len(sentiment_df)):
        total_count += features[i]
      
      print("#","!","?","@")
      print('total count ' + sentiment + " " + intensity +":\n",total_count)
charCounterByIntensity()



----anger----
# ! ? @
total count anger high:
 [163  78  23  92]
# ! ? @
total count anger medium:
 [120  30  27 136]
# ! ? @
total count anger low:
 [186  31  15  86]

----sadness----
# ! ? @
total count sadness high:
 [221  35  18  73]
# ! ? @
total count sadness medium:
 [149  32  23 124]
# ! ? @
total count sadness low:
 [250  62  26 118]

----fear----
# ! ? @
total count fear high:
 [397  67  39 121]
# ! ? @
total count fear medium:
 [357  73  40 195]
# ! ? @
total count fear low:
 [323  83  34 176]

----joy----
# ! ? @
total count joy high:
 [303 124   4 144]
# ! ? @
total count joy medium:
 [196  43  12 115]
# ! ? @
total count joy low:
 [ 70  19  17 113]


In [ ]:
class CharsCountTransformer2(BaseEstimator, TransformerMixin):
    def get_relevant_chars(self, tweet):
        num_hashtags = tweet.count('#')
        num_exclamations = tweet.count('!')
        num_interrogations = tweet.count('?')
        num_at = tweet.count('@')
        num_ast = tweet.count('*')
        num_sus = tweet.count('...')
        return [num_hashtags, num_exclamations, num_interrogations, num_at, num_ast, num_sus]

    def transform(self, X, y=None):
        chars = []
        for tweet in X:
            chars.append(self.get_relevant_chars(tweet))

        return np.array(chars)

    def fit(self, X, y=None):
        return self

In [ ]:
class ElongatedWordsCountTransformer(BaseEstimator, TransformerMixin):
    def get_relevant_chars(self, tweet):
        regex = re.compile(r"([a-z])\1{2}")
        num_elon = 0
        for word in tweet.split():
            if regex.search(word):
              num_elon += 1
        return [num_elon]

    def transform(self, X, y=None):
        words = []
        for tweet in X:
            words.append(self.get_relevant_chars(tweet))

        return np.array(words)

    def fit(self, X, y=None):
        return self

In [ ]:
sentiments = ['anger','sadness','fear','joy']
sentiment_intensities = ['high','medium','low']
smallest_set = {'anger':161,'sadness':197,'fear':270,'joy':195}

def elonCounterByIntensity():
  for sentiment in sentiments:
    print("\n----"+sentiment+"----")
    data_frame = train[sentiment]
    for intensity in sentiment_intensities:
      sentiment_df = data_frame[data_frame['sentiment_intensity']==intensity].sample(smallest_set[sentiment])
      features = ElongatedWordsCountTransformer().transform(sentiment_df.tweet)
      total_count = [0]
      for i in range(len(sentiment_df)):
        total_count += features[i]
      
      print("#","!","?","@")
      print('total count ' + sentiment + " " + intensity +":\n",total_count)
elonCounterByIntensity()


----anger----
# ! ? @
total count anger high:
 [3]
# ! ? @
total count anger medium:
 [4]
# ! ? @
total count anger low:
 [11]

----sadness----
# ! ? @
total count sadness high:
 [2]
# ! ? @
total count sadness medium:
 [7]
# ! ? @
total count sadness low:
 [5]

----fear----
# ! ? @
total count fear high:
 [3]
# ! ? @
total count fear medium:
 [4]
# ! ? @
total count fear low:
 [3]

----joy----
# ! ? @
total count joy high:
 [2]
# ! ? @
total count joy medium:
 [3]
# ! ? @
total count joy low:
 [3]


In [ ]:
class LexiconCountTransformer(BaseEstimator, TransformerMixin):
    lexicons = {
        'anger':['fuming','angry','people','fucking','fuck','irritate'],
        'sadness':['depression','sad','sadness','unhappy','amp','grim'],
        'fear':['nervous','nightmare','anxiety','panic','terror','im'],
        'joy':['happy','love','smile','hilarious','joyful','great']
    }
    lexicon_list = []

    def set_lexicon_list(self, sentiment):
        self.lexicon_list = self.lexicons[sentiment]
        return self

    def get_relevant_words(self, tweet):
        results = [0]*len(self.lexicon_list)
        for i in range(len(self.lexicon_list)):
            cv = CountVectorizer()
            cv_fit = cv.fit_transform([tweet])
            df = pd.DataFrame({'token':cv.get_feature_names_out(), 'count':cv_fit.toarray().sum(axis=0)})
            present = df[df['token']==self.lexicon_list[i]]
            if len(present) > 0:
                results[i] = int(df[df['token']==self.lexicon_list[i]]['count'])**2
            else:
                results[i] = 0
        return results

    def transform(self, X, y=None):
        words = []
        for tweet in X:
            words.append(self.get_relevant_words(tweet))

        return np.array(words)

    def fit(self, X, y=None):
        return self

In [ ]:
# Veamos que sucede si ejecutamos el transformer
sample = train['anger'].sample(5).tweet
sample_features = LexiconCountTransformer().set_lexicon_list('anger').transform(sample)

# Se puede verificar que el conteo de símbolos es consistente con el transformer creado.
print(f'Tweet original: {sample.iloc[0]}')
print(f'Features creados: {sample_features[0]}')

Tweet original: jamming out to fury in math class wanting to die hbu
Features creados: [0 0 0 0 0 0]


### Definir la representación y el clasificador

Para esto, definiremos Pipelines. Un `Pipeline` es una lista de transformaciones y un estimador(clasificador) ubicado al final el cual define el flujo que seguiran nuestros datos dentro del sistema que creemos. Nos permite ejecutar facilmente el mismo proceso sobre todos los datasets que usemos, simplificando así nuestra programación.

El pipeline más básico que podemos hacer es transformar el dataset a Bag of Words y después usar clasificar el BoW usando NaiveBayes:

```python
    Pipeline([('bow', CountVectorizer()), ('clf', MultinomialNB())])
```


Ahora, si queremos usar nuestra transformación para agregar las features que creamos, usaremos `FeatureUnion`. Esta simplemente concatenará los vectores resultantes de ejecutar BoW y los Transformer en un solo vector.

```python
    Pipeline([('features',FeatureUnion([('bow', CountVectorizer()),
                                        ('chars_count',CharsCountTransformer())])),
              ('clf', MultinomialNB())])

```





Recuerden que cada pipeline representa un sistema de clasificación distinto. Por lo mismo, deben instanciar uno por cada problema que resuelvan. De lo contrario, podrían solapar resultados.  Para esto, les recomendamos crear los pipeline en distintas funciones, como la siguiente:

In [ ]:
from sklearn.feature_extraction.text import TfidfTransformer
def get_experiment_0_pipeline():

    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer()),
                                    ])), ('clf', MultinomialNB())])

In [ ]:
from sklearn.feature_extraction.text import TfidfTransformer
def get_experiment_1F_pipeline():

    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer2()),
                                    ])), ('clf', MultinomialNB())])

In [ ]:
from sklearn.feature_extraction.text import TfidfTransformer
def get_experiment_2F_pipeline():

    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer2()),
                                    ('elon_words_count', ElongatedWordsCountTransformer()),
                                    ])), ('clf', MultinomialNB())])

In [ ]:
from sklearn.feature_extraction.text import TfidfTransformer
def get_experiment_3F_pipelines():
    pipelines = {}
    pipelines['anger'] = Pipeline([('features',
                          FeatureUnion([('bow', CountVectorizer()),
                                        ('chars_count', CharsCountTransformer2()),
                                        ('lexicon_count', LexiconCountTransformer().set_lexicon_list('anger')),
                                        ])), ('clf', MultinomialNB())])
    pipelines['sadness'] = Pipeline([('features',
                          FeatureUnion([('bow', CountVectorizer()),
                                        ('chars_count', CharsCountTransformer2()),
                                        ('lexicon_count', LexiconCountTransformer().set_lexicon_list('sadness')),
                                        ])), ('clf', MultinomialNB())])
    pipelines['joy'] = Pipeline([('features',
                          FeatureUnion([('bow', CountVectorizer()),
                                        ('chars_count', CharsCountTransformer2()),
                                        ('lexicon_count', LexiconCountTransformer().set_lexicon_list('joy')),
                                        ])), ('clf', MultinomialNB())])
    pipelines['fear'] = Pipeline([('features',
                          FeatureUnion([('bow', CountVectorizer()),
                                        ('chars_count', CharsCountTransformer2()),
                                        ('lexicon_count', LexiconCountTransformer().set_lexicon_list('fear')),
                                        ])), ('clf', MultinomialNB())])
    return pipelines

### Ejecutar el pipeline para algún dataset

In [ ]:
def run(dataset, dataset_name, pipeline):
    """Creamos el pipeline y luego lo ejecutamos el pipeline sobre un dataset. 
    Retorna el modelo ya entrenado mas sus labels asociadas y los scores obtenidos al evaluarlo."""

    # Dividimos el dataset en train y test, aún no se transforma de Strings a valores numéricos.
    X_train, X_test, y_train, y_test = train_test_split(
        dataset.tweet,
        dataset.sentiment_intensity,
        shuffle=True,
        test_size=0.33)
    
    print(f'# Datos de entrenamiento en dataset {dataset_name}: {len(X_train)}')
    print(f'# Datos de testing en dataset {dataset_name}: {len(X_test)}')

    # Entrenamos el clasificador (Ejecuta el entrenamiento sobre todo el pipeline). 
    # En este caso el Bag of Words es el encargado de transformar de Strings a vectores numéricos.
    pipeline.fit(X_train, y_train)

    # Predecimos las probabilidades de intensidad de cada elemento del set de prueba.
    predicted_probabilities = pipeline.predict_proba(X_test)

    # Obtenemos el orden de las clases aprendidas.
    learned_labels = pipeline.classes_
    
    # Evaluamos:
    scores = evaluate(predicted_probabilities, y_test, learned_labels, dataset_name)
    return pipeline, learned_labels, scores

### Ejecutar el sistema creado por cada train set

Este código crea y entrena los 4 sistemas de clasificación y luego los evalua. Para los experimentos, pueden copiar este código variando el pipeline cuantas veces estimen conveniente.

In [ ]:
classifiers = []
learned_labels_array = []
scores_array = []

# Por cada nombre_dataset, dataset en train ('anger', 'fear', 'joy', 'sadness')
for dataset_name, dataset in train.items():
  
    # creamos el pipeline
    pipeline = get_experiment_0_pipeline()
    
    # ejecutamos el pipeline sobre el dataset
    classifier, learned_labels, scores = run(dataset, dataset_name, pipeline)

    # guardamos el clasificador entrenado (en realidad es el pipeline ya entrenado...)
    classifiers.append(classifier)

    # guardamos las labels aprendidas por el clasificador
    learned_labels_array.append(learned_labels)

    # guardamos los scores obtenidos
    scores_array.append(scores)

# print avg scores
print(
    "Average scores:\n\n",
    "Average AUC: {0:.3g}\t Average Kappa: {1:.3g}\t Average Accuracy: {2:.3g}"
    .format(*np.array(scores_array).mean(axis=0)))

# Datos de entrenamiento en dataset anger: 630
# Datos de testing en dataset anger: 311
Confusion Matrix for anger:

[[  8  46   0]
 [  6 186   7]
 [  0  47  11]]

Classification Report:

              precision    recall  f1-score   support

         low       0.57      0.15      0.24        54
      medium       0.67      0.93      0.78       199
        high       0.61      0.19      0.29        58

    accuracy                           0.66       311
   macro avg       0.62      0.42      0.43       311
weighted avg       0.64      0.66      0.59       311

Scores:

AUC:  0.63	Kappa: 0.163	Accuracy: 0.659
------------------------------------------------------

# Datos de entrenamiento en dataset fear: 842
# Datos de testing en dataset fear: 415
Confusion Matrix for fear:

[[ 18  74   2]
 [ 13 208  15]
 [  1  57  27]]

Classification Report:

              precision    recall  f1-score   support

         low       0.56      0.19      0.29        94
      medium       0.61      0.8

In [ ]:
classifiers = []
learned_labels_array = []
scores_array = []

# Por cada nombre_dataset, dataset en train ('anger', 'fear', 'joy', 'sadness')
for dataset_name, dataset in train.items():
  
    # creamos el pipeline
    pipeline = get_experiment_1_pipeline()
    
    # ejecutamos el pipeline sobre el dataset
    classifier, learned_labels, scores = run(dataset, dataset_name, pipeline)

    # guardamos el clasificador entrenado (en realidad es el pipeline ya entrenado...)
    classifiers.append(classifier)

    # guardamos las labels aprendidas por el clasificador
    learned_labels_array.append(learned_labels)

    # guardamos los scores obtenidos
    scores_array.append(scores)

# print avg scores
print(
    "Average scores:\n\n",
    "Average AUC: {0:.3g}\t Average Kappa: {1:.3g}\t Average Accuracy: {2:.3g}"
    .format(*np.array(scores_array).mean(axis=0)))

# Datos de entrenamiento en dataset anger: 630
# Datos de testing en dataset anger: 311
Confusion Matrix for anger:

[[  6  37   0]
 [  8 201   8]
 [  1  39  11]]

Classification Report:

              precision    recall  f1-score   support

         low       0.40      0.14      0.21        43
      medium       0.73      0.93      0.81       217
        high       0.58      0.22      0.31        51

    accuracy                           0.70       311
   macro avg       0.57      0.43      0.44       311
weighted avg       0.66      0.70      0.65       311

Scores:

AUC:  0.594	Kappa: 0.174	Accuracy: 0.701
------------------------------------------------------

# Datos de entrenamiento en dataset fear: 842
# Datos de testing en dataset fear: 415
Confusion Matrix for fear:

[[ 18  64   1]
 [ 16 209  17]
 [  1  68  21]]

Classification Report:

              precision    recall  f1-score   support

         low       0.51      0.22      0.31        83
      medium       0.61      0.

In [ ]:
classifiers = []
learned_labels_array = []
scores_array = []

# Por cada nombre_dataset, dataset en train ('anger', 'fear', 'joy', 'sadness')
for dataset_name, dataset in train.items():
  
    # creamos el pipeline
    pipeline = get_experiment_2_pipeline()
    
    # ejecutamos el pipeline sobre el dataset
    classifier, learned_labels, scores = run(dataset, dataset_name, pipeline)

    # guardamos el clasificador entrenado (en realidad es el pipeline ya entrenado...)
    classifiers.append(classifier)

    # guardamos las labels aprendidas por el clasificador
    learned_labels_array.append(learned_labels)

    # guardamos los scores obtenidos
    scores_array.append(scores)

# print avg scores
print(
    "Average scores:\n\n",
    "Average AUC: {0:.3g}\t Average Kappa: {1:.3g}\t Average Accuracy: {2:.3g}"
    .format(*np.array(scores_array).mean(axis=0)))

# Datos de entrenamiento en dataset anger: 630
# Datos de testing en dataset anger: 311
Confusion Matrix for anger:

[[  6  43   0]
 [  8 197   7]
 [  0  46   4]]

Classification Report:

              precision    recall  f1-score   support

         low       0.43      0.12      0.19        49
      medium       0.69      0.93      0.79       212
        high       0.36      0.08      0.13        50

    accuracy                           0.67       311
   macro avg       0.49      0.38      0.37       311
weighted avg       0.60      0.67      0.59       311

Scores:

AUC:  0.612	Kappa: 0.072	Accuracy: 0.666
------------------------------------------------------

# Datos de entrenamiento en dataset fear: 842
# Datos de testing en dataset fear: 415
Confusion Matrix for fear:

[[ 15  73   2]
 [ 13 198  16]
 [  3  68  27]]

Classification Report:

              precision    recall  f1-score   support

         low       0.48      0.17      0.25        90
      medium       0.58      0.

In [ ]:
classifiers = []
learned_labels_array = []
scores_array = []

# Por cada nombre_dataset, dataset en train ('anger', 'fear', 'joy', 'sadness')
for dataset_name, dataset in train.items():
  
    # creamos el pipeline
    pipeline = get_experiment_3_pipelines()[dataset_name]
    
    # ejecutamos el pipeline sobre el dataset
    classifier, learned_labels, scores = run(dataset, dataset_name, pipeline)

    # guardamos el clasificador entrenado (en realidad es el pipeline ya entrenado...)
    classifiers.append(classifier)

    # guardamos las labels aprendidas por el clasificador
    learned_labels_array.append(learned_labels)

    # guardamos los scores obtenidos
    scores_array.append(scores)

# print avg scores
print(
    "Average scores:\n\n",
    "Average AUC: {0:.3g}\t Average Kappa: {1:.3g}\t Average Accuracy: {2:.3g}"
    .format(*np.array(scores_array).mean(axis=0)))

# Datos de entrenamiento en dataset anger: 630
# Datos de testing en dataset anger: 311
Confusion Matrix for anger:

[[  7  34   0]
 [  7 194   9]
 [  0  52   8]]

Classification Report:

              precision    recall  f1-score   support

         low       0.50      0.17      0.25        41
      medium       0.69      0.92      0.79       210
        high       0.47      0.13      0.21        60

    accuracy                           0.67       311
   macro avg       0.55      0.41      0.42       311
weighted avg       0.62      0.67      0.61       311

Scores:

AUC:  0.602	Kappa: 0.127	Accuracy: 0.672
------------------------------------------------------

# Datos de entrenamiento en dataset fear: 842
# Datos de testing en dataset fear: 415
Confusion Matrix for fear:

[[ 17  70   5]
 [ 18 198  16]
 [  1  47  43]]

Classification Report:

              precision    recall  f1-score   support

         low       0.47      0.18      0.27        92
      medium       0.63      0.

### Predecir los target set y crear la submission

Aquí predecimos los target set usando los clasificadores creados y creamos los archivos de las submissions.

In [ ]:
def predict_target(dataset, classifier, labels):
    # Predecir las probabilidades de intensidad de cada elemento del target set.
    predicted = pd.DataFrame(classifier.predict_proba(dataset.tweet), columns=labels)
    
    # Agregar ids
    predicted['id'] = dataset.id.values
    # Reordenar las columnas
    predicted = predicted[['id', 'low', 'medium', 'high']]
    return predicted

In [ ]:
predicted_target = {}

# Crear carpeta ./predictions
if (not os.path.exists('./predictions')):
    os.mkdir('./predictions')

else:
    # Eliminar predicciones anteriores:
    shutil.rmtree('./predictions')
    os.mkdir('./predictions')

# por cada target set:
for idx, key in enumerate(target):
    # Predecirlo
    predicted_target[key] = predict_target(target[key], classifiers[idx],
                                           learned_labels_array[idx])
    # Guardar predicciones en archivos separados. 
    predicted_target[key].to_csv('./predictions/{}-pred.txt'.format(key),
                                 sep='\t',
                                 header=False,
                                 index=False)

# Crear archivo zip
a = shutil.make_archive('predictions', 'zip', './predictions')

## **7. Resultados**

    Muestre y analice sus resultados aquí

A continuación una tabla de ejemplo para mostrar los resultados.  En este caso sólo está el experimento del baseline más otro clasificador, pero ustedes debiesen generar una mayor cantidad de experimentos. 

| No. | Approach                       || Dataset   | AUC   | Kappa | Accuracy |
|-----|--------------------------------||-----------|-------|-------|----------|
|     | Features        | Clasifier     |           |       |       |          |
| 0   | bow+chars_count | MultinomialNB | anger     | 0.622 | 0.163 | 0.688    |
|     |                 |               | fear      | 0.597 | 0.091 | 0.559    |
|     |                 |               | joy       | 0.728 | 0.251 | 0.601    |
|     |                 |               | sadness   | 0.645 | 0.166 | 0.581    |
|     |                 |               |**average**| 0.648 | 0.168 | 0.607    |
| 1   | tus features    | tu clasifier  | anger     |       |       |          |
|     |                 |               | fear      |       |       |          |
|     |                 |               | joy       |       |       |          |
|     |                 |               | sadness   |       |       |          |
|     |                 |               |**average**|       |       |          |


## **8. Conclusiones**

    Escriba aquí sus conclusiones